scripts/showcase/samplesheet-prep-test.py

```diff
@@ -1,55 +1,399 @@
+            IDT Index , unless stated
+            Baymax run#
+            Comments
+            rRNA
+            qPCR ID
+            Sample_ID (SampleSheet)
+            ==========                ==============================================================
+    """
+
+    tracking_sheets = []
+
+    # Initialise in a excel file
+    logger.info("Loading excel file")
+    xl = pd.ExcelFile(tracking_sheet)
+
+    # Check sheet names - we just want from 2019 onwards
+    valid_sheet_names = [sheet_name
+                         for sheet_name in xl.sheet_names
+                         if sheet_name.isnumeric()
+                         and sheet_name not in OMITTED_YEAR_SHEETS]
+    if len(valid_sheet_names) == 0:
+        logger.error("Could not find any valid sheet names in {}".format(tracking_sheet))
+        sys.exit(1)
+
+    # read tracking sheet
+    for sheet_name in valid_sheet_names:
+        logger.info("Reading sheet %s" % sheet_name)
+        tracking_sheets.append(xl.parse(header=0, sheet_name=sheet_name))
+
+    tracking_sheet_df = pd.concat(tracking_sheets)
+
+    return tracking_sheet_df
+
+
+def add_sample_type_to_sample_sheet(tracking_sheet_df, sample_sheet_df):
+    """
+    Truncate the tracking_sheet_df to just the Sample_ID and the type, then run a merge on the two dataframes
+    """
+
+    # Get the list of samples in the sample sheet
+    sample_list = sample_sheet_df['Sample_ID'].unique().tolist()
+
+    # A bit going on here:
+    # 1. Slim down the tracking data frame to just the Sample ID and the type of sample.
+    # 2. Rename the columns to 'Sample_ID and Sample_Type (the latter is used in the TSO500 sample sheet)
+    # 3. Filter to rows only present in this sample sheet
+    # 4. Add the column 'Type' that we then group by. TSO_DNA and TSO_RNA are set to TSO
+    slimmed_tracking_df = tracking_sheet_df[['Type', 'Sample_ID (SampleSheet)']].\
+        rename(columns={"Sample_ID (SampleSheet)": "Sample_ID",
+                        "Type": "Sample_Type"}).\
+        query("Sample_ID in @sample_list")
+
+    slimmed_tracking_df['Type'] = slimmed_tracking_df['Sample_Type'].apply(lambda x: 'TSO' if x.startswith("TSO") else x)
+
+    # Merge and return the data frames with the 'Type included
+    return pd.merge(sample_sheet_df, slimmed_tracking_df, on='Sample_ID', how='left')
+
+
+def write_sample_sheet(sample_sheet_df, sample_sheet_header_rows, sample_sheet_output_file):
+    """
+    Given a sample type, a full sample sheet df, and a header row, write out the sample,sheet as a csv file
+    """
+
+    with open(sample_sheet_output_file, 'w') as sample_sheet_output_h:
+        # Write out the header rows
+        sample_sheet_output_h.writelines(sample_sheet_header_rows)
+        # Write out [Data]
+        sample_sheet_output_h.write("{}\n".format(HEADER_LINE_PRECURSOR))
+        # Write out sample sheet
+        sample_sheet_df.\
+            to_csv(sample_sheet_output_h, sep=",", header=True, index=False)
```

Comment:
Please merge this with the `write_sample_sheets` function

Reply:
https://github.com/umccr/infrastructure/pull/16/files#diff-9803ecbbdf15c05af6e26ccc75f88c84R354-R378

Reply:
Done! Preferably we'd have separate functions for the modification process and the write out process. There's a bit going on in those four lines at the bottom so having its own function with a doc string could make the code easier to read. Maybe a change in name from `write_sample_sheets` to `output_sample_sheets` and `write_sample_sheet` to `write_sample_sheet_to_file` would have reduced ambiguity around the function naming.